src/security.js

```diff
@@ -0,0 +1,326 @@
+
+const BEARER_OFFSET = 7;
+
+/**
+ * @typedef {Object} AuthenticationRequirement
+ * @property {Array<string>} [allowedRoles]
+ * @property {Array<string>} [allowedPermissions]
+ */
+
+/**
+ * @typedef {Object} TokenRequest
+ * @property {string} spaceId the space for which the token will be generated
+ * @property {Array<string>} roleKeys the role keys for which the token should be generated
+ * @property {string} generator provides attribution for the key in the description
+ * @property {number} [expiresInMinutes] the number of minutes before the token expires (or 14 days)
+ */
+
+/**
+ * @typedef {Object} TokenPayload
+ * @property {Array<string>} permissions the permissions granted to this token
+ * @property {Array<string>} roles the permissions granted to this token
+ * @property {string} sub the subject (or identifier) for this token
+ * @property {string} tenantId the primary tenant for the token
+ * @property {Array<string>} [tenantIds] the optional list of additional
+ *  allowed tenants for the token
+ * @property {string} type the type of the token
+ * @property {number} exp the timestamp at which the token will expire
+ */
+
+export class Security {
+  /**
+   * Checks to see if any of the allowed permissions are present in the actual permissions,
+   * using globbing expansion
+   * @param {Array<string>} allowedPermissions the list of permissions which are allowed
+   * @param {Array<string>} actualPermissions the actual permissions from the request
+   * @returns {boolean}
+   */
+  static #hasPermissions(allowedPermissions, actualPermissions) {
+    return allowedPermissions
+      .some((required) => actualPermissions.some((granted) => minimatch(required, granted)));
+  }
+
+  /**
+   * @type {string}
+   */
+  #apiHost;
+
+  #certificate;
+
+  /**
+   * @type {import('./common-typedefs.js').Logger}
+   */
+  #log;
+
+  /**
+   * Maps the role names to role IDs
+   * @type {Record<string,string>}
+   */
+  #roles;
+
+  /**
+   * @type {SecretsManager}
+   */
+  #secretsManager;
+
+  constructor(context) {
+    const helper = new ContextHelper(context);
+    this.#apiHost = helper.getEnv().SECURITY_API_HOST || 'https://api.frontegg.com';
+    this.#log = helper.getLog();
+    this.#secretsManager = context.secretsManager
+      || new SecretsManager({
+        ...helper.extractAwsConfig(),
+        application: 'frontegg',
+        scope: context.scope,
+      });
+  }
+
+  /**
+   * Authenticates that the request is allowed to continue based on
+   * the authentication requirements
+   * @param {AuthenticationRequirement} [authentication] the requirements
+   *    for authentication to succeed
+   * @param {Request} req the request
+   * @returns {Promise<void>}
+   */
+  async authenticate(req, authentication) {
+    const token = this.#getToken(req);
+    const spaceId = this.#getRequiredHeader(req, 'x-space-id');
+    const payload = jwt.decode(token);
+
+    // first validate that the tenant is correct
```

Comment:
qs: is there a way to validate the company id is a valid company id too/first?

Reply:
for the record, I am working on getting the company id (parent tenant id) into the JWT. this is possible using the jwt generation [frontegg prehook](https://docs.frontegg.com/docs/subscribing-to-prehooks). adding this new hook over in [adobe/content-lake-frontegg-service](https://github.com/adobe/content-lake-frontegg-service).